### PR TITLE
simplification in substitution and function evaluation

### PR DESCRIPTION
**Pull request by andreistefanescu**

### CODE REVIEW

**Comment by andreistefanescu** [comment]
@yilongli what do you think?

File: java-backend/src/main/java/org/kframework/backend/java/kil/Term.java [hunk 1]
```diff
@@ -131,17 +131,11 @@ public Term substituteWithBinders(Map<Variable, ? extends Term> substitution, Te
...
      */
     public Term substituteAndEvaluate(Map<Variable, ? extends Term> substitution, TermContext context) {
-        // TODO(AndreiS): disable the check below when proving things until this is properly fixed by Cosmin
-        if (context.definition().context().krunOptions == null
-                || context.definition().context().krunOptions.experimental.prove == null) {
-            // TODO(AndreiS): assert that there are not any unevaluated functions in this term
-            if (substitution.isEmpty() || isGround()) {
-                return this;
-            }
+        if ((!substitution.isEmpty() && !isGround()) || !isNormal()) {
```
Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/kil/Term.java]: This condition is little hard to read. I would prefer `return isNormal && (substitution.isEmpty() || isGround()) ? this : (Term) this.accept(new SubstituteAndEvaluateTransformer(substitution, context));`

File: java-backend/src/main/java/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java [hunk 2]
```diff
@@ -41,19 +41,7 @@ public SubstituteAndEvaluateTransformer(Map<Variable, ? extends Term> substituti
...
-            set1 = set2;
-            set2 = tmp;
-        }
-        for (Variable variable : set1) {
-            if (set2.contains(variable)) {
-                return true;
-            }
-        }
-        return false;
+        return ((!substitution.isEmpty() && !object.isGround()) || !object.isNormal());
```
Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java]: The original version is more precise?

Comment (andreistefanescu) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java]: Yes it is, but I think it's too expensive.

Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java]: But it also avoids unnecessary recursion. So does the new version show speedup on any example?

Comment (andreistefanescu) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java]: Well, for example the avl delete goes from 18s to 17.5. I know it's not much, but it helps a little.

Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/symbolic/SubstituteAndEvaluateTransformer.java]: ok, then I am fine with it. BTW, can you remove the redundant parentheses around the expressions?

**Comment by dwightguth** [comment]
@andreistefanescu you didn't actually address Yilong's comments before you merged. Please perform a more careful reading of this pull request and then recreate the diff.

**Comment by dwightguth** [comment]
I'm sorry, I misread this diff. Never mind. I'll re-merge the change.

**Pull request merged**